### PR TITLE
Make application/julia a text MIME

### PR DESCRIPTION
**Pull request by TotalVerb**
Add `application/julia` as a text MIME, and sort the list of text MIMEs.

I wonder if `istextmime` could be made a little more intelligent. All mime types with `text` should be text MIMEs, right?

### CODE REVIEW

**Comment by stevengj** [comment]
`istextmime` should return true for anything that should be represented by a `String` (as opposed to raw binary data), and as far as I can tell this is true for anything in `text/` ... indeed, according to [RFC6557](https://tools.ietf.org/html/rfc6657), which seems to be routinely ignored today thanks to UTF-8, the default charset of `text/*` data is ASCII.

(I personally hate the fact that programming languages are supposedly in the "application" namespace and not in "text".   But IANA does not seem particularly consistent here.... see [text/ecmascript](http://www.iana.org/assignments/media-types/text/ecmascript) vs [application/javascript](http://www.iana.org/assignments/media-types/application/javascript), and those are basically the same language!)

**Comment by TotalVerb** [comment]
Note that `text/ecmascript` is marked OBSOLETE while [application/ecmascript](http://www.iana.org/assignments/media-types/application/ecmascript) is COMMON.

**Comment by TotalVerb** [comment]
I suspect this definition would work as well:

```julia
istextmime{T}(::MIME{T}) = startswith(string(T), "text/") | T in [... list of other mimes ...]
```

**Comment by TotalVerb** [comment]
The `@textmime` macro can be retained in the event a package needs to monkey-patch a text MIME in, although I don't think this is considered a public API.

**Comment by nalimilan** [comment]
FWIW, the [Freedesktop.org shared MIME database](https://cgit.freedesktop.org/xdg/shared-mime-info/tree/freedesktop.org.xml.in) has a `<sub-class-of type="text/plain"/>` attribute identifying text files. Can be useful as a reference, in particular if we want to ensure before each release that we haven't missed new MIME types (e.g. for new languages).

File: base/multimedia.jl [hunk 1]
```diff
@@ -81,7 +81,7 @@ istextmime(m::AbstractString) = istextmime(MIME(m))
...
+for mime in ["application/atom+xml", "application/ecmascript", "application/javascript", "application/julia", "application/json", "application/postscript", "application/rdf+xml", "application/rss+xml", "application/x-latex", "application/xhtml+xml", "application/xml", "application/xml-dtd", "image/svg+xml", "model/vrml", "model/x3d+vrml", "model/x3d+xml", "text/calendar", "text/cmd", "text/css", "text/csv", "text/html", "text/javascript", "text/latex", "text/markdown", "text/n3", "text/plain", "text/richtext", "text/sgml", "text/tab-separated-values", "text/vcard", "text/vnd.graphviz", "text/x-setext", "text/x-vcalendar", "text/x-vcard", "text/xml"]
```
Comment (tkelman) [inline base/multimedia.jl]: should line wrap this while we're at it

**Comment by TotalVerb** [comment]
There seems to be demand for a more advanced change. I'm playing around with turning `istextmime` into a trait, and then dispatching on this trait instead of the clunky `@textmime` macro. Unfortunately, returning a `Bool` is currently not good enough (this is essentially https://github.com/JuliaLang/julia/issues/17880), because the return type is known to be a constant too late, and the type instability in `reprmime` cannot be removed (though LLVM can remove the dead code).

So the new change implements `istextmime` using a lower-level `mimetypetype` trait (name could/should be changed to something more descriptive, but I couldn't figure out what). It turns out that this actually simplifies the code quite a bit at this point. While I was working on this, I moved docstrings over from `helpdb.jl`. In fact, these docstrings mostly duplicated the big comment near the top about text vs. binary mimes.

Remaining issues:

- Do I need to deprecate `@textmime` (as I've done here), or should I just rip it out? There [exist](https://github.com/kmsquire/XClipboard.jl/blob/987daebf4634ee3be19857ac677decb3eda2786a/src/xmime.jl) [usages](https://github.com/Keno/TexExtensions.jl/blob/7cd22de764d49bd0faf942d74f9126b66e0ddf0f/src/TexExtensions.jl) in the wild.
- What (if anything) should be exported?

File: base/multimedia.jl (hunk 1 above)
Comment (TotalVerb) [inline base/multimedia.jl]: Is there a standard way to wrap long `for` loops? I just pushed to the beginning of line here.

**Comment by TotalVerb** [comment]
Hmm, it looks like the trait did not avoid the `::Union{Array{UInt8,1},String}` pessimization of `reprmime`'s return type. I guess the implementation of `istextmime` might be too complicated. Will try a `@pure` annotation.

**Comment by TotalVerb** [comment]
The compiler really doesn't like the complexity of the `mimetypetype` function; adding `Base.@pure` did not seem to help. My suspicion is that `stringmime` and `reprmime` are fairly infrequently used in performance-critical situations, and that when they are the cost of allocation would dominate the cost of the type instability. Perhaps this performance issue can be tolerated until https://github.com/JuliaLang/julia/issues/14324 can be implemented.

Alternatively, we can put the `text/` mime types back into the list to maintain.

File: base/multimedia.jl (hunk 1 above)
Comment (tkelman) [inline base/multimedia.jl]: I personally follow the matlab convention of indenting line continuations of block openers one level deeper than the body of the block

File: base/multimedia.jl [hunk 2]
```diff
@@ -33,56 +33,89 @@ mimewritable{mime}(::MIME{mime}, x) =
...
+requested `mime` type. This is similar to [`reprmime`](:func:`reprmime`) except
+that binary data is base64-encoded as an ASCII string.
+"""
+stringmime(m::MIME, x) = stringmime(mimetypetype(m), m, x)
+stringmime(::IsText, m::MIME, x) = reprmime(m, x)
+stringmime(::IsBytes, m::MIME, x) = base64encode(verbose_show, m, x)
+stringmime(::IsBytes, m::MIME, x::Vector{UInt8}) = base64encode(write, x)
+
 macro textmime(mime)
+    Base.depwarn(string("`@textmime mime` is deprecated; use ",
```
Comment (tkelman) [inline base/multimedia.jl]: if deprecated, it should probably move to `deprecated.jl`

Comment (TotalVerb) [inline base/multimedia.jl]: It needs to be in the `Multimedia` module though.

Comment (tkelman) [inline base/multimedia.jl]: ah, is it not exported from there? could be eval'ed into that module

Comment (TotalVerb) [inline base/multimedia.jl]: Done.

**Comment by TotalVerb** [comment]
I doubt the performance is an issue, so I think this should be good to go now?

**Comment by stevengj** [comment]
I guess so.  I have to be honest that I'm not a huge fan of the whole traits technique in cases like this; it seems like an overly complicated way to determine something at compile time that is unlikely to ever be performance-sensitive.

**Comment by TotalVerb** [comment]
@stevengj Well, due to inference limitations these traits aren't even working properly. (We need #14324 to remove the type instability with traits, and with booleans we need #17880). I'm happy to do this with booleans if people like that better.

**Comment by TotalVerb** [comment]
In fact, the original approach was with booleans; then I switched to traits because I ran into #17880, and then I discovered that the functions was too complex for even traits to fix the type instability...

**Comment by stevengj** [comment]
Speaking for myself, I'd prefer booleans.

**Comment by TotalVerb** [comment]
@stevengj Just pushed a boolean version. The boolean version is far simpler too, and probably also has lower compilation overhead. I like it also.

**Comment by TotalVerb** [comment]
Oops, I need to fix the deprecated `@textmime` macro too.

**Comment by stevengj** [comment]
I don't think we generally add tests for deprecations.

File: base/deprecated.jl [hunk 3]
```diff
@@ -999,4 +999,14 @@ macro vectorize_2arg(S,f)
 end
 export @vectorize_1arg, @vectorize_2arg
 
+# Deprecate @textmime into the Multimedia module, #18441
+eval(Multimedia, :(macro textmime(mime)
+    Base.depwarn(string("`@textmime mime` is deprecated; use ",
+        "`Base.Multimedia.mimetypetype(::MIME{mime}) = ",
```
Comment (stevengj) [inline base/deprecated.jl]: Shouldn't the deprecation now be `Base.istextmime(::MIME{mime}) = true`?

Comment (TotalVerb) [inline base/deprecated.jl]: Oops. The code is correct, but I forgot to update the text.

Comment (TotalVerb) [inline base/deprecated.jl]: Fixed.

**Comment by TotalVerb** [comment]
Is this good to squash and merge?

**Pull request merged**